services/hierarchy/src/main/kotlin/PackageService.kt

```diff
@@ -0,0 +1,341 @@
+ * License-Filename: LICENSE
+ */
+
+package org.eclipse.apoapsis.ortserver.services
+
+import org.eclipse.apoapsis.ortserver.dao.dbQuery
+import org.eclipse.apoapsis.ortserver.dao.tables.AnalyzerJobsTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.AnalyzerRunsTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.AuthorsTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.MappedDeclaredLicensesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.PackageDao
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.PackagesAnalyzerRunsTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.PackagesAuthorsTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.PackagesDeclaredLicensesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.PackagesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.ProcessedDeclaredLicensesMappedDeclaredLicensesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.ProcessedDeclaredLicensesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.ProcessedDeclaredLicensesUnmappedDeclaredLicensesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.analyzer.UnmappedDeclaredLicensesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.shared.DeclaredLicensesTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.shared.IdentifiersTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.shared.RemoteArtifactsTable
+import org.eclipse.apoapsis.ortserver.dao.tables.runs.shared.VcsInfoTable
+import org.eclipse.apoapsis.ortserver.dao.utils.listCustomQuery
+import org.eclipse.apoapsis.ortserver.model.RepositoryType
+import org.eclipse.apoapsis.ortserver.model.runs.Identifier
+import org.eclipse.apoapsis.ortserver.model.runs.Package
+import org.eclipse.apoapsis.ortserver.model.runs.ProcessedDeclaredLicense
+import org.eclipse.apoapsis.ortserver.model.runs.RemoteArtifact
+import org.eclipse.apoapsis.ortserver.model.runs.VcsInfo
+import org.eclipse.apoapsis.ortserver.model.util.ListQueryParameters
+import org.eclipse.apoapsis.ortserver.model.util.ListQueryResult
+
+import org.jetbrains.exposed.dao.id.EntityID
+import org.jetbrains.exposed.sql.Database
+import org.jetbrains.exposed.sql.JoinType
+import org.jetbrains.exposed.sql.ResultRow
+import org.jetbrains.exposed.sql.alias
+
+/**
+ * Aliases needed for the query and by the mapping function
+ */
+val binaryArtifactTable = RemoteArtifactsTable.alias("ra1")
+val sourceArtifactTable = RemoteArtifactsTable.alias("ra2")
+val vcsTable = VcsInfoTable.alias("vcs1")
+val vcsProcessedTable = VcsInfoTable.alias("vcs2")
+
+/**
+ * A service to interact with packages.
+ */
+class PackageService(private val db: Database) {
+    suspend fun listForOrtRunId(
+        ortRunId: Long,
+        parameters: ListQueryParameters = ListQueryParameters.DEFAULT
+    ): ListQueryResult<Package> = db.dbQuery {
```

Comment:
I haven't tested it, but I think the implementation could be simplified by using the DAO API:

```kotlin
val query = PackagesTable
    .innerJoin(PackagesAnalyzerRunsTable)
    .innerJoin(AnalyzerRunsTable)
    .innerJoin(AnalyzerJobsTable)
    .select(PackagesTable.columns)
    .where { AnalyzerJobsTable.ortRunId eq ortRunId }
    .withDistinct()

val packages = PackageDao.wrapRows(query).map { it.mapToModel() }
```

The `mapToModel()` function already has the code to load all related data.

The only thing missing here is applying the pagination and sorting to the query. The code for that you could take from the `SizedIterable<T>.apply` helper function in `utils/Extensions.kt` in the `dao` module.

This is probably a little less efficient as, for example, the data from the `remote_artifacts` table is loaded in separate queries, but as long as the `limit' is not very large this should not matter (and you already do it for the authors and processed declared license anyway).

Now that I think of it, the code for the vulnerabilities endpoint could probably be simplified in the same way.

Reply:
I have applied this approach to the `VulnerabilityService` in #1078 and could remove a lot of code.

Reply:
Ok good thanks, I'll try similar changes here.